src/storage/petastorm_storage_engine.py

```diff
@@ -0,0 +1,141 @@
+from src.storage.abstract_storage_engine import AbstractStorageEngine
+from src.utils.logging_manager import LoggingLevel
+from src.utils.logging_manager import LoggingManager
+from src.configuration.configuration_manager import ConfigurationManager
+
+from petastorm.unischema import dict_to_spark_row
+from petastorm.predicates import in_set, in_lambda
+from petastorm import make_reader
+from typing import Iterator, Dict
+
+
+class PetastormStorageEngine(AbstractStorageEngine):
+
+    def __init__(self):
+        """
+        Maintain a long live spark session and context.
+        """
+        self._spark = Session()
+        self.spark_session = self._spark.get_session()
+        self.spark_context = self._spark.get_context()
+
+    def _spark_url(self, table: DataFrameMetadata) -> str:
+        """
+        Generate a spark/petastorm url given a table
+        """
+        eva_dir = ConfigurationManager().get_value("core", "location")
+        output_url = os.path.join(eva_dir, table.name)
+
+        return output_url
+
+    def create(self, table: DataFrameMetadata):
+        """
+        Create an empty dataframe in petastorm.
+        """
+        empty_rdd = self.spark_context.emptyRDD()
+
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            self.spark_session.createDataFrame(empty_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('overwrite') \
+                .parquet(self._spark_url(table))
+
+
+    def write_row(self, table: DataFrameMetadata, rows: []):
+        """
+        Write rows into the dataframe.
+
+        Arguments:
+            rows List[Dict{key, value}]: We can formally define a dataType Row.
+            keys within the Dict should be consistent with the table.schema.
+        """
+        def row_generator(x):
+            return rows[x]
+
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            rows_rdd = self.spark_context.parallelize(range(len(rows)))\
+                .map(row_generator)\
+                .map(lambda x: dict_to_spark_row(table.schema.petastorm_schema, x))
+
+            self.spark_session.createDataFrame(rows_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('append') \
+                .parquet(self._spark_url(table))
+
+    def read(self, table: DataFrameMetadata) -> Iterator[Dict]:
+        """
+        Read the dataframe.
+
+        Return:
+            Iterator of Dict. Each Dict represents a Row.
+        """
+        with make_reader(self._spark_url(table)) as reader:
```

Comment:
I have written the Reader for this job. How about we let the petastorm reader handle this?
I think you can let it be the same for now. We can update it after my changes are merged.

Reply:
Agree

Reply:
If you push reader (or maybe you already pushed), I can do that now.